viewer/packages/utilities/src/events/InputHandler.ts

```diff
@@ -0,0 +1,120 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+import { clickOrTouchEventOffset } from './clickOrTouchEventOffset';
+import { EventTrigger } from './EventTrigger';
+import debounce from 'lodash/debounce';
+import { assertNever } from '@reveal/utilities';
+import { Vector2 } from 'three';
+
+type PointerEventDelegate = (event: { offsetX: number; offsetY: number }) => void;
+
+export class InputHandler {
+  private readonly domElement: HTMLElement;
+
+  private readonly _events = {
+    click: new EventTrigger<PointerEventDelegate>(),
+    hover: new EventTrigger<PointerEventDelegate>()
+  };
+
+  private setupEventListeners() {
+    const { domElement } = this;
+
+    const maxMoveDistance = 8;
+    const maxClickDuration = 250;
+
+    let pointerDown = false;
+    let pointerDownTimestamp = 0;
+    let validClick = false;
+
+    const startOffset = new Vector2();
+
+    const onHoverCallback = debounce((e: MouseEvent) => {
+      this._events.hover.fire(clickOrTouchEventOffset(e, domElement));
```

Comment:
Move this implementation and other event handler implementations below to member functions to make `setupEventListeners` smaller

Reply:
Others require the context of this function so moved only parts of them but looks better I think. Not sure about number of arguments, though...